_posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc

```diff
@@ -0,0 +1,2045 @@
+  infrastructure with a few clicks in the AWS UI. The problem with out-of-band changes is that it cause configuration
+  drift and leads to complicated bugs.
+
+- *There is only one environment for the code*. Unlike with application code where you can easily have many copies of the
+  code running at the same time, having duplicate environments of infrastructure code is very expensive. It is expensive
+  as is having additional copies of your infrastructure for dev, stage, and prod, imagine having three copies of that
+  triplet for infrastructure (e.g `infra-dev-dev`, `infra-dev-stage`, `infra-dev-prod` and so on). This not only makes
+  testing the infrastructure code hard, but also limits branching strategies: you don't want any long lived branches,
+  and you don't want to deploy any infrastructure from feature branches to avoid state conflicts.
+
+With these differences in mind, let's discuss what the workflow for infrastructure code looks like.
+
+[[repository_organization]]
+==== Repository organization
+
+Just like with application code, it is important to use version control for infrastructure code. Version control is
+critical for being able to track changes and implementing review processes for your code. However, unlike with
+application code, you will typically want at least two separate version control repositories for your infrastructure
+code: one for infrastructure modules (typically called `infrastructure-modules` or just `modules`), and one for your live
+environment configuration (typically called `infrastructure-live` or just `live`).
+
+modules::
+  This repository should contain reusable infrastructure code to deploy common components of your infrastructure. Think
+  of this repo as the "blueprints" that define the way your company configures infrastructure. For example, you might
+  define an infrastructure module for deploying RDS databases.
+
+live::
+  This repository should contain the live configuration of your infrastructure for each of your environments. If
+  `modules` contain "blueprints" then this repo contain the "houses" that were built using the "blueprints." For
+  example, you might define an environment that contains multiple RDS databases, each configured with different
+  parameter options (e.g name, server size, disk configuration, etc).
+
+Since the nature of the release artifacts for the two repositories are different, naturally the CI/CD pipeline for the
+repositories will also be different.
+
+For `modules`, release artifacts will be versioned immutable snapshots of the code that you can consume in the `live`
+configuration. The versioned artifacts ensure that you get a known configuration that won't change overtime (unless you
+request a different version). This helps facilitate consistent roll out of the particular infrastructure component
+across multiple environments.
+
+In contrast, for `live`, there is typically no release artifact. Instead, you would apply the code to the cloud to mark
```

Comment:
This makes sense. I was consciously trying to avoid repeating myself often, but when the writing was stretched across multiple days, naturally regress to the habit words.